tests/api.rs

```diff
@@ -226,6 +226,73 @@ impl scroll::ctx::FromCtx<scroll::Endian> for Bar {
         }
     }
 }
+#[derive(Debug)]
+struct StrDrop<'a> {
+    #[allow(unused)]
+    s: &'a str,
+}
+
+impl<'a> scroll::ctx::TryFromCtx<'a, scroll::Endian> for StrDrop<'a> {
+    type Error = scroll::Error;
+    fn try_from_ctx(
+        bytes: &'a [u8],
+        _ctx: scroll::Endian,
+    ) -> ::std::result::Result<(Self, usize), Self::Error> {
+        let mut offset = 0;
+        let s = bytes.gread(&mut offset)?;
+        Ok((Self { s }, offset))
+    }
+}
+
+impl<'a> Drop for StrDrop<'a> {
+    fn drop(&mut self) {
+        println!("Dropping {self:?}");
+    }
+}
+
+#[derive(Debug)]
+struct StringDrop {
+    #[allow(unused)]
+    s: String,
+}
+
+impl<'a> scroll::ctx::TryFromCtx<'a, scroll::Endian> for StringDrop {
+    type Error = scroll::Error;
+    fn try_from_ctx(
+        bytes: &'a [u8],
+        _ctx: scroll::Endian,
+    ) -> ::std::result::Result<(Self, usize), Self::Error> {
+        let mut offset = 0;
+        let s: &str = bytes.gread(&mut offset)?;
+        Ok((Self { s: String::from(s) }, offset))
+    }
+}
+
+impl Drop for StringDrop {
+    fn drop(&mut self) {
+        println!("Dropping {self:?}");
+    }
+}
+
+#[test]
+#[should_panic]
+fn test_fixed_array_str() {
+    use scroll::Pread;
+    let bytes = [0x45, 0x42, 0x0, 0x45, 0x41];
+    let _res = bytes
+        .pread_with::<[StrDrop; 3]>(0, Default::default())
```

Comment:
i need to understand what's going on here, i'd expect this to panic with an array of 2, since we don't have null terminating byte, and that's the str pread default ctx iirc

Reply:
Both of these tests do panic, but since that is the expected behavior, I added `#[should_panic]` to the test.

Reply:
But their length is 3.  I’d expect them to panic at 2. Which they do not

Reply:
Ah I see what you mean, I'm currently on mobile, but will look into this in an hour or so.

Reply:
Ok, I looked into it and ran a test, with this:
```rust
#[test]
fn string_test() {
    use scroll::Pread;
    let bytes = [0x45, 0x42, 0x0, 0x45, 0x41];
    let mut offset = 0;
    bytes.gread_with::<&str>(&mut offset, StrCtx::Delimiter(0)).unwrap();
    bytes.gread_with::<&str>(&mut offset, StrCtx::Delimiter(0)).unwrap();
}
```
This is an issue with the `TryFromCtx` impl on `&str`, it doesn't return an error if no null bytes is encountered, but instead creates a string from as many bytes as are present. This is unrelated to this PR and needs to be fixed separately.

Reply:
I opened a PR fixing this #96.

Reply:
I don't think we should block this PR on that issue; adding generics is useful enough, and the only thing that's weird is the test; you can just remove the str based tests, though it would be useful to have an allocated, non fixed size/context dependent type (String was for this); I suggest this:

```rust
type MyCtx = ();

#[derive(Debug)]
struct BytesDrop<'a> {
    #[allow(unused)]
    s: &'a [u8],
}

impl<'a> scroll::ctx::TryFromCtx<'a, MyCtx> for BytesDrop<'a> {
    type Error = scroll::Error;
    fn try_from_ctx(
        src: &'a [u8],
        _ctx: MyCtx,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let len = src.iter().take_while(|c| **c != 0x0).count();
        if len > src.len() {
            return Err(scroll::Error::TooBig {
                size: len,
                len: src.len(),
            });
        };
        Ok((Self { s: &src[..len] }, len))
    }
}

impl<'a> Drop for BytesDrop<'a> {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
    }
}

#[derive(Debug)]
struct VecDrop {
    #[allow(unused)]
    s: Vec<u8>,
}

impl<'a> scroll::ctx::TryFromCtx<'a, MyCtx> for VecDrop {
    type Error = scroll::Error;
    fn try_from_ctx(
        bytes: &'a [u8],
        _ctx: MyCtx,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        let s: BytesDrop = bytes.gread(&mut offset)?;
        Ok((Self { s: Vec::from(s.s) }, offset))
    }
}

impl Drop for VecDrop {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
        drop(&mut self.s);
    }
}

#[test]
fn test_fixed_array_str() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41]);
    let res = bytes.pread_with::<[BytesDrop; 2]>(0, Default::default());
    println!("{res:?}");
    assert!(res.is_ok());
}

#[test]
fn test_fixed_array_string() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41, 0x0]);
    let res = bytes.pread_with::<[VecDrop; 2]>(0, Default::default());
    println!("{res:?}");
    assert!(res.is_ok());
}
```

can add check that the inner bytes of the BytesDrop and VecDrop are as expected, if want. After that, let's merge this!

Reply:
I just pushed this. Blocking this will probably not be necessary.